File: frontend/packages/technicalEvaluationMethod.js
import {
  EVALUATION_METHOD_CODES,
  isCombinedEvaluationMethod,
  normalizeEvaluationMethod,
  parseTechnicalScore,
} from "./evaluationMethodRules.js";
import { parseEvaluationMetadataForDisplay } from "./evaluationMetadata.js";

export { parseTechnicalScore } from "./evaluationMethodRules.js";

export const TECHNICAL_EVALUATION_METHODS = Object.freeze({
  PASS_FAIL: "pass_fail",
  SCORE: "score",
});

const FORCED_PASS_FAIL_FORMS = new Set([
  "chao hang canh tranh",
  "chi dinh thau",
  "chi dinh thau rut gon",
  "lua chon nha thau trong truong hop dac biet",
]);

function normalize(value) {
  return String(value ?? "")
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "")
    .toLocaleLowerCase("vi")
    .replace(/đ/g, "d")
    .replace(/[^a-z0-9]+/g, " ")
    .trim();
}

export function normalizeTechnicalEvaluationMethod(value) {
  const token = normalize(value);
  if (["score", "scoring", "cham diem", "diem"].includes(token)) {
    return TECHNICAL_EVALUATION_METHODS.SCORE;
  }
  if (["pass fail", "pass_fail", "dat khong dat", "dat va khong dat"].includes(token)) {
    return TECHNICAL_EVALUATION_METHODS.PASS_FAIL;
  }
  return "";
}

export function isCombinedTechnicalPriceMethod(valueOrPackage = {}) {
  return isCombinedEvaluationMethod(valueOrPackage);
}

function metadataBlock(pkg, roundType) {
  const metadata = parseEvaluationMetadataForDisplay(
    pkg?.danhGiaHsdtMetadata,
  ).metadata;
  return roundType === "single" ? metadata : metadata?.[roundType] || {};
}

export function getStoredTechnicalEvaluationMethod(pkg = {}, roundType = "single") {
  const block = metadataBlock(pkg, roundType);
  return normalizeTechnicalEvaluationMethod(
    block.technicalEvaluationMethod
      || block.phuongPhapDanhGiaKyThuat
      || pkg.technicalEvaluationMethod
      || pkg.phuongPhapDanhGiaKyThuat,
  );
}

export function getForcedTechnicalEvaluationMethod(pkg = {}) {
  const field = normalize(pkg.linhVuc || pkg.loaiGoiThau || pkg.loaiGoi || pkg.category);
  if (field === "tu van" || field.startsWith("tu van ")) {
    return TECHNICAL_EVALUATION_METHODS.SCORE;
  }
  if (FORCED_PASS_FAIL_FORMS.has(normalize(pkg.hinhThucLuaChon))) {
    return TECHNICAL_EVALUATION_METHODS.PASS_FAIL;
  }
  const overallMethod = normalizeEvaluationMethod(pkg);
  if ([
    EVALUATION_METHOD_CODES.COMBINED_TECHNICAL_PRICE,
    EVALUATION_METHOD_CODES.TECHNICAL_BASED,
  ].includes(overallMethod)) {
    return TECHNICAL_EVALUATION_METHODS.SCORE;
  }
  return "";
}

function inferImportedMethod(criteria = []) {
  const imported = criteria.filter((criterion) => (
    criterion.group === "technical"
    && ["muasamcong", "custom"].includes(String(criterion.source || "").trim().toLowerCase())
  ));
  if (imported.length === 0) return "";
  return imported.some((criterion) => criterion.resultType === "score")
    ? TECHNICAL_EVALUATION_METHODS.SCORE
    : TECHNICAL_EVALUATION_METHODS.PASS_FAIL;
}

export function resolveTechnicalEvaluationMethod({
  pkg = {},
  roundType = "single",
  report = null,
  criteria = [],
  draftMethod = "",
} = {}) {
  return getForcedTechnicalEvaluationMethod(pkg)
    || normalizeTechnicalEvaluationMethod(draftMethod)
    || getStoredTechnicalEvaluationMethod(pkg, roundType)
    || normalizeTechnicalEvaluationMethod(
      report?.extension?.technicalEvaluationMethod
        || report?.extension?.phuongPhapDanhGiaKyThuat,
    )
    || inferImportedMethod(criteria);
}

export function requiresTechnicalScore(pkg = {}, roundType = "single") {
  return resolveTechnicalEvaluationMethod({ pkg, roundType })
    === TECHNICAL_EVALUATION_METHODS.SCORE;
}

export function configureBidTechnicalScoreInputs(root, pkg = {}, roundType = "single") {
  const scoreRequired = requiresTechnicalScore(pkg, roundType);
  root?.querySelectorAll?.("input.mt-dg-ky-thuat").forEach((input) => {
    if (!scoreRequired) {
      input.removeAttribute?.("data-technical-score-required");
      input.removeAttribute?.("required");
      input.removeAttribute?.("aria-required");
      return;
    }
    const current = String(input.value ?? "").trim();
    const parsed = parseTechnicalScore(current);
    const hadInvalidLegacyValue = Boolean(current) && parsed === null;
    input.type = "number";
    input.setAttribute?.("min", "0");
    input.setAttribute?.("step", "any");
    input.setAttribute?.("inputmode", "decimal");
    input.setAttribute?.("required", "true");
    input.setAttribute?.("aria-required", "true");
    input.setAttribute?.("data-technical-score-required", "true");
    input.placeholder = "Nhập điểm kỹ thuật...";
    if (current) input.value = parsed === null ? "" : String(parsed);
    if (
      hadInvalidLegacyValue
      && typeof input.dispatchEvent === "function"
      && typeof globalThis.Event === "function"
    ) {
      input.dispatchEvent(new globalThis.Event("input", { bubbles: true }));
    }
  });
  return scoreRequired;
}

export function applyTechnicalEvaluationMethod(criteria = [], method = "") {
  const normalizedMethod = normalizeTechnicalEvaluationMethod(method);
  if (!normalizedMethod) return criteria;
  return criteria.map((criterion) => criterion.group === "technical"
    ? { ...criterion, resultType: normalizedMethod === TECHNICAL_EVALUATION_METHODS.SCORE ? "score" : "pass_fail" }
    : criterion);
}

File: backend/documents/technical_evaluation_method.py
"""Resolve the technical evaluation method used by evaluation Excel workbooks."""

from __future__ import annotations

import re
import unicodedata
from collections.abc import Mapping

from backend.sync.evaluation_metadata import parse_evaluation_metadata


PASS_FAIL = "pass_fail"
SCORE = "score"
UNKNOWN = "unknown"

_FORCED_PASS_FAIL_FORMS = {
    "chao hang canh tranh",
    "chi dinh thau",
    "chi dinh thau rut gon",
    "lua chon nha thau trong truong hop dac biet",
}
_SCORE_METHODS = {
    "combined technical price",
    "ket hop giua ky thuat va gia",
    "ket hop ky thuat va gia",
    "technical based",
    "dua tren ky thuat",
}


def _normalize(value):
    text = unicodedata.normalize("NFD", str(value or ""))
    text = "".join(char for char in text if not unicodedata.combining(char))
    return re.sub(r"[^a-z0-9]+", " ", text.lower().replace("đ", "d")).strip()


def _first(package, *keys):
    for key in keys:
        value = package.get(key)
        if value not in (None, ""):
            return value
    return ""


def _explicit_method(value):
    explicit = _normalize(value)
    if explicit in {"score", "scoring", "cham diem", "diem"}:
        return SCORE
    if explicit in {"pass fail", "pass_fail", "dat khong dat", "dat va khong dat"}:
        return PASS_FAIL
    return ""


def _stored_metadata_block(package, round_type):
    raw_metadata = _first(package, "danhGiaHsdtMetadata", "danh_gia_hsdt_metadata")
    try:
        metadata = parse_evaluation_metadata(raw_metadata, require_version=False)
    except ValueError:
        # Match the frontend display resolver: invalid historical metadata must
        # not invent an evaluation method.
        return {}
    if round_type == "single":
        return metadata
    block = metadata.get(round_type)
    return dict(block) if isinstance(block, Mapping) else {}


def _stored_technical_evaluation_method(package, round_type):
    block = _stored_metadata_block(package, round_type)
    return _explicit_method(_first(
        block,
        "technicalEvaluationMethod",
        "technical_evaluation_method",
        "phuongPhapDanhGiaKyThuat",
        "phuong_phap_danh_gia_ky_thuat",
    )) or _explicit_method(_first(
        package,
        "technicalEvaluationMethod",
        "technical_evaluation_method",
        "phuongPhapDanhGiaKyThuat",
        "phuong_phap_danh_gia_ky_thuat",
    ))


def _forced_technical_evaluation_method(package):
    field = _normalize(_first(
        package,
        "linhVuc", "linh_vuc", "loaiGoiThau", "loai_goi_thau",
        "loaiGoi", "loai_goi", "category",
    ))
    if field == "tu van" or field.startswith("tu van "):
        return SCORE

    selection_form = _normalize(_first(
        package, "hinhThucLuaChon", "hinh_thuc_lua_chon",
    ))
    if selection_form in _FORCED_PASS_FAIL_FORMS:
        return PASS_FAIL

    evaluation_method = _normalize(_first(
        package,
        "phuongPhapDanhGia", "phuong_phap_danh_gia", "evaluationMethodCode",
    ))
    if evaluation_method in _SCORE_METHODS:
        return SCORE
    return ""


def resolve_technical_evaluation_method(package_context, *, round_type="single"):
    """Return ``pass_fail``, ``score``, or ``unknown`` using frontend semantics.

    ``round_type`` selects persisted round metadata for a 1G2T package.  Domain
    rules determined from the package remain authoritative, as they do in the
    frontend, before a stored technical-method preference is considered.
    """

    package = package_context if isinstance(package_context, Mapping) else {}
    normalized_round = str(round_type or "single").strip().lower()
    if normalized_round not in {"single", "technical", "financial"}:
        normalized_round = "single"
    return (
        _forced_technical_evaluation_method(package)
        or _stored_technical_evaluation_method(package, normalized_round)
        or UNKNOWN
    )
